### PR TITLE
Exact DSL draft

### PR DESCRIPTION
**Pull request by ustits**
Usage example can be found in `NotBlankTrimmedString` in test package.

- Decided to keep only one abstraction `Exact`. It both keeps a constraint and helper functions for creating "exact" type
- Chose `from` instead of `create` because, imho, it sounds more natural, e.g. `NotBlankString.fromOrThrow(str)` versus `NotBlankString.createOrThrow(str)`. Don't have a strong opinion here so it can be changed
- Made `Exact` an interface otherwise the clients won't be able to extend from other classes if needed
- There are two ways to define an `Exact`: interface implementation and `exact()` method call
- `Exact` can be combined via `and` function like `NotBlankString and TrimmedString`

### CODE REVIEW

File: src/commonMain/kotlin/arrow.exact/Exact.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  fun fromOrEither(value: A): Either<ExactError, B>
+
+  fun fromOrNull(value: A): B? {
+    return fromOrEither(value).orNull()
+  }
+
+  fun fromOrThrow(value: A): B {
+    return when (val result = fromOrEither(value)) {
+      is Either.Left -> throw IllegalArgumentException(result.value.message)
```
Comment (ILIYANGERMANOV) [inline src/commonMain/kotlin/arrow.exact/Exact.kt:15]: We should probably create an `ArrowExactException` containing more information but that's not critical for the draft. I've defined an example for such error in #5 .

File: src/commonMain/kotlin/arrow.exact/Exact.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package arrow.exact
+
+import arrow.core.*
+
+interface Exact<A, out B> {
+
+  fun fromOrEither(value: A): Either<ExactError, B>
```
[the remark this answers is not in the document]
Comment (ILIYANGERMANOV) [reply src/commonMain/kotlin/arrow.exact/Exact.kt:7]: Wouldn't it be better to use raise as the method that must be implemented. 
```kotlin
context(Raise<ExactError>)
fun from(value: A): B
```

or without Context receivers:
```kotlin
fun Raise<ExactError>.from(value: A): B
```

File: src/commonMain/kotlin/arrow.exact/ExactError.kt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+package arrow.exact
+
+class ExactError(val message: String)
```
Comment (ILIYANGERMANOV) [inline src/commonMain/kotlin/arrow.exact/ExactError.kt:3]: We can make this class `open` so people can define custom exact errros that can be handled easily programmatically.

File: src/commonTest/kotlin/arrow/exact/Strings.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package arrow.exact
+
+import arrow.core.Either
+import arrow.core.right
+
+class NotBlankString private constructor(val str: String) {
+  companion object : Exact<String, NotBlankString> by exact(String::isNotBlank, ::NotBlankString)
+}
+
+class NotBlankTrimmedString private constructor(val str: String) {
```
Comment (ILIYANGERMANOV) [inline src/commonTest/kotlin/arrow/exact/Strings.kt:10]: With my proposed changes this can look like:
```kotlin
class NotBlankTrimmedString private constructor(val str: String) {
  companion object : Exact<String, NotBlankTrimmedString> by exact(::NotBlankTrimmedString) {
     val trimmed = it.trim()
     ensure(trimmed::isNotBlank) { ExactError("blank string") }
     trimmed
  }
}
```

OR

```kotlin
class NotBlankTrimmedString private constructor(val str: String) {
  companion object : Exact<String, NotBlankTrimmedString> by trimmed() and exact(::NotBlankTrimmedString, String::isNotBlank)
}
```

**Review by ILIYANGERMANOV — COMMENTED** [review]
Thank you for taking the time to initiate the work @ustits! Overall, great design - I like it 👍 
Added a couple of suggestions around:
- reducing boilerplate 
- Arrow.`Raise`
- naming improvements 

Please,  let me know what you think. The goal is discussion and perfection through iterations and feedback.

**Comment by ILIYANGERMANOV** [comment]
💡 In a nutshell, we have two types of operations:
- `validation`: a predicate `(A) -> Boolean` that is:
  - safe to compose
  - commutative, composition order doesn't matter
  - e.g `NotBlankString`
- `transformation`: mapping `(A) -> B`:
  - order matters
  - e.g. `TrimmedString`

**TL;DR;**
- A type-safe Exact builder pattern 
- Makes it clear that the order of operations matter

This is an overkill but just sharing for brainstorming purposes.
I'm thinking about creating an additional Exact DSL which makes building complex types easier similarly to the "builder" pattern.
```kotlin
companion object : Exact<String, CustomerId> by exact()
     .mustBe(String::isNotBlank) // predicate
     .transform(String::trim) // transformation
     .transformOrFail { // both
        when {
            it.startsWith("u/") -> User(it.drop(2))
            it.endsWith("@arrow"l -> Admin(it)
            else -> fail("Invalid id: $it")
        }
      }
     .predicate { ... }
     .transform { it.uppercased() }
     .build(::CustomerId)
```

**Comment by ILIYANGERMANOV** [comment]
For visibility mentioning #4

**Review by nomisRev — APPROVED** [review]
Really nice work @ustits 🙌 Awesome to get a community effort going here 😍

I left some nits mostly, and some thoughts but I like everything I saw here.

**Comment by nomisRev** [comment]
@ustits I invited you to the repo, so you can also create branches directly not his repo.

File: src/commonMain/kotlin/arrow.exact/Exact.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ustits) [reply src/commonMain/kotlin/arrow.exact/Exact.kt:7]: Totally agree with `from`. I think it will be the main method to be used, so it is better not to pollute it with extra words

File: src/commonMain/kotlin/arrow.exact/Exact.kt (hunk 1 above)
Comment (ustits) [reply src/commonMain/kotlin/arrow.exact/Exact.kt:15]: @ILIYANGERMANOV You are right, will create a custom exception and extend it from `IllegalArgumentException` as @nomisRev has supposed. Indeed it will be useful to put there some extra information about the failed validation

But now I see that as a client I probably would want to define my own exception, maybe we can design it like in Result.getOrElse:

```
fromOrElse(onFailure: (kotlin.Throwable) -> R): R
```

File: src/commonMain/kotlin/arrow.exact/ExactDsl.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  private val exact2: Exact<B, C>
+) : Exact<A, C> {
+
+  override fun fromOrEither(value: A): Either<ExactError, C> {
+    return exact1.fromOrEither(value)
+      .flatMap { exact2.fromOrEither(it) }
+  }
+}
+
+fun <A, B> exact(predicate: Predicate<A>, constructor: (A) -> B): Exact<A, B> {
```
[the remark this answers is not in the document]
Comment (ustits) [reply src/commonMain/kotlin/arrow.exact/ExactDsl.kt:16]: @ILIYANGERMANOV Unfortunately we can't implement it in such a way, kotlin compiler thinks that after `{` there must be a class body, not a lambda, so eventually there is no difference in arguments order. On the other hand `exact` can be used not only in delegation but also as regular method call:

``` kotlin
fun main() {
  exact(::NotBlankString) {
    it.isNotBlank()
  }
}
```

But it is rather synthetic example, doesn't look like a real use case

@nomisRev I am not quite familiar with Raise api, can you tell me where to look for implementation examples?

[the remark this answers is not in the document]
Comment (ILIYANGERMANOV) [reply src/commonMain/kotlin/arrow.exact/ExactDsl.kt:16]: Thank you for the explanation @ustits! I did all the code examples from my Fold w/o an actual IDE/compiler - apologies for that! Regarding `Raise` - I'd recommend:
- https://arrow-kt.io/learn/typed-errors/working-with-typed-errors/
- https://github.com/arrow-kt/arrow/pull/3038 (I've put some KDoc of Raise with short examples if you're looking for a TL;DR;) 

**Raise TL;DR;**
- like `throws` but type-safe, in order to raise you need a `Raise<A>` context
- to "catch" raise use `recover`
- interoperable with `Either`
- @nomisRev can give you a better explanation and details but that should be enough to get you started

**Review by ILIYANGERMANOV — APPROVED** [review]
The design is already good! I think we can merge it and build on top of it. @ustits feel free to merge :+1: Great work! :clap: 

_After merging this one, we can start creating small PRs with minor improvements and have discussions inside them._

File: src/commonMain/kotlin/arrow.exact/Exact.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ILIYANGERMANOV) [reply src/commonMain/kotlin/arrow.exact/Exact.kt:7]: Let's just rename it to `from` and resolve this conversion. Seems like we all agree that `from` is a good name.

File: src/commonMain/kotlin/arrow.exact/Exact.kt (hunk 1 above)
Comment (ILIYANGERMANOV) [reply src/commonMain/kotlin/arrow.exact/Exact.kt]: Perfect, I imagine it like that - custom exception extending `IllegalArgumentException` :+1:

File: src/commonMain/kotlin/arrow.exact/ExactDsl.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (ustits) [reply src/commonMain/kotlin/arrow.exact/ExactDsl.kt:16]: Thanks @ILIYANGERMANOV , can you tell me, am I moving in the right direction? As I understood I need to use `recover` on the constraint and map the result onto `Either`. Btw in such scenario probably we don't need a constructor as an argument:

``` kotlin
fun <A, B> exact(constraint: Raise<ExactError>.(A) -> B): Exact<A, B> {
  return object : Exact<A, B> {
    override fun from(value: A): Either<ExactError, B> {
      return recover({
        constraint(value).right()
      }) {
        error -> error.left()
      }
    }
  }
}
```

[the remark this answers is not in the document]
Comment (ILIYANGERMANOV) [reply src/commonMain/kotlin/arrow.exact/ExactDsl.kt:16]: @ustits yes, it looks good to me! Thank you for making the change :)
A valid example usage would look like this, right?
```kotlin
companion object : Exact<String, NotBlankTrimmedString> by exact({
   ensureNotNull(it.trim().takeIf(String::isNotBlank)?.let(::NotBlankTrimmedString)) { 
       ExactError("Cannot be blank!")
   }
})
```

```kotlin
companion object : Exact<Float, Percent> by exact({
   Percent(it.coerceIn(0f, 1f))
})
```

I like this API a lot! It's a lot simpler when we get rid of the constructor.
The constructor can be required only for the predicate `exact` case.

[the remark this answers is not in the document]
Comment (nomisRev) [reply src/commonMain/kotlin/arrow.exact/ExactDsl.kt:16]: @ustits instead of using `recover` with `left` and `right` you can also just use the `either` block. It works through `Raise` 😉. TL;DR `Raise` is a higher-level abstraction that can work over errors of `E`. So for `Either` it's just `E`, but for `Option` it's for example `None` or `typealias Null = Nothing?` for _nullable types`. 

```kotlin
      return either { constraint(value) }
```

**Comment by nomisRev** [comment]
Agreed, let's merge this PR 💪 @ustits or @ILIYANGERMANOV can you confirm that you can merge this PR with the proper approvals? Just to know I gave you the correct permissions to the repo 😅

**Comment by ILIYANGERMANOV** [comment]
@nomisRev @ustits I merged it - we have the correct permissions. Wanted to merge it directly because I want to play around and build on top of it - it's more convenient to have it merged instead of doing stacked PRs.

**Comment by ustits** [comment]
@ILIYANGERMANOV @nomisRev Thanks for the help! 

I haven't included changes with Raise, will do a separate PR for this

**Comment by nomisRev** [comment]
Thank you so much @ustits & @ILIYANGERMANOV! 🙌 Super exciting to see this taking shape 😍 🥳

File: src/commonMain/kotlin/arrow.exact/Exact.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  fun fromOrThrow(value: A): B {
+    return when (val result = from(value)) {
+      is Either.Left -> throw ExactException(result.value.message)
+      is Either.Right -> result.value
+    }
+  }
+}
+
+open class ExactError(val message: String)
```
Comment (CLOVIS-AI) [inline src/commonMain/kotlin/arrow.exact/Exact.kt:21]: Why an open class and not an interface?

Comment (ILIYANGERMANOV) [reply src/commonMain/kotlin/arrow.exact/Exact.kt:21]: @CLOVIS-AI That's outdated. Have a look at #8 👈 and #13 👈

Comment (CLOVIS-AI) [reply src/commonMain/kotlin/arrow.exact/Exact.kt:21]: Thanks for the link :)
